### PR TITLE
Compare View: TR-55 Runoff Charts

### PR DESCRIPTION
**Pull request by kellyi**
## Overview

This PR adds the TR-55 Compare View charts and sets up the table & chart values to adjust accordingly. gif demo's below.

Connects #2072 

### Demo

compare-charts

### Notes

Currently there are a few quirks & limitations with this implementation, which we can either handle here or spin off elsewhere. 

1. I did not implement the popovers, which seem to be complicated enough to handle in another card.  NVD3 has tooltips by default, but since the desired behavior's that the hovering over one value on the bar chart opens corresponding popovers for all three, which isn't a built-in NVD3 feature. I think we may end up wanting to use Bootstrap popovers for this.

2. The bars themselves match the width of the scenario title placeholders when there are few enough of them to display in the scenario header without overflow. However, when there are more, the chart div seems not to expand in size beyond what's visible so the bars get squished.

3. There's currently some flicker as new charts are shown when retrieving any of the results, which means it renders once per scenario if the precipitation slider value changes. Not sure what is the best way to handle this -- I think we may want to set some value to check whether any of the scenarios is currently polling and show a loading state if not?

4. I think the maximum y axis value for "Combined Hydrology" should be the value the precipitation slider has been set to, but it seemed like the value these scenarios had was slightly off the number displayed for the slider. Not sure what is the best way to handle this, but we'll need to pass it in to the chart to set the `yDomain`.

Happy to try to address these here or to spin them off elsewhere; figured I'd put up the PR for review rather than spinning wheels on it.

## Testing
- get this branch, then ~`update` and `server`~ `bundle.sh` & hard refresh in the browser
- create a new TR-55 project then create a few scenarios including some which would cause varied values for runoff/infiltration (think: veg basin gswi conservation practice or high density land cover)
- open the compare view and verify that you see the new charts and that the values seem accurate to the underlying data
- move the precipitation slider and verify that when the polling stops, the values in the chart have changed
- click to see the table view, then move the precipitation slider, and verify that those values change too

### CODE REVIEW

**Comment by kellyi** [comment]
Realized I left off a change which forces the y axis for the non-Combined-Hydrology charts to go slightly beyond the max value of the bars:

screen shot 2017-08-09 at 5 51 28 pm

Amended the commit & pushed it anew.

**Comment by rajadain** [comment]
> get this branch, then `update` and `server`

This is one of our few projects to which this doesn't apply.

**Comment by kellyi** [comment]
> This is one of our few projects to which this doesn't apply.

Whoops! `bundle.sh` then hard refresh instead!

**Comment by rajadain** [comment]
The tick labels on the Combined Hydrology chart are incorrect:

image

This should be 10 instead of 1. The labels on the other charts are correct.

File: src/mmw/js/src/compare/views.js [hunk 1]
```diff
@@ -46,6 +48,18 @@ var CompareWindow2 = Marionette.LayoutView.extend({
         sectionsRegion: '.compare-sections',
     },
 
+    initialize: function() {
+        var self = this;
+        this.model.get('scenarios').forEach(function(scenario) {
+            scenario.get('results').on('change', function(s) {
+                if (s.attributes.name === 'runoff' && !s.attributes.polling) {
```
Comment (arottersman) [inline src/mmw/js/src/compare/views.js]: We should still use the backbone syntax: `s.get('name')` and `s.get('polling')`. We also may want to refer to it as `r`, for "result"

File: src/mmw/js/src/compare/views.js [hunk 2]
```diff
@@ -472,8 +534,74 @@ function getTr55Tabs(scenarios) {
                 })
             },
         ],
-        // TODO Make Runoff charts
-        runoffCharts = [],
+        runoffCharts = [
```
Comment (arottersman) [inline src/mmw/js/src/compare/views.js]: @rajadain Was there a reason for all the arrays/objects in this function to not be actual models?
 Why doesn't this return a`TabsCollection`?

**Comment by kellyi** [comment]
> The tick labels on the Combined Hydrology chart are incorrect:

Good catch. The combined hydrology chart needs to get the precipitation slider value as an input so it can use it to set its `yDomain` max value.

**Comment by rajadain** [comment]
This is interesting: when the precipitation value is changed, we run a TR-55 job for all four scenarios (in this example), and update the chart as the values return. As the initial values return, there is great variation in the chart, as we have some new low values coexisting with high old values. Once all the new ones are gathered, the chart rescales itself to the new standard. This is not very obvious to the user (especially because there are no spinners or loading symbols, more on that later), but the effect can be jarring:



Can we make it so that the chart updates when _every_ scenario has finished running instead of individual ones?

**Comment by rajadain** [comment]
In the wireframes: https://marvelapp.com/8hi9jg9/screen/29125398, the background color of the columns alternates. Was any thought given to this here?

**Comment by rajadain** [comment]
It looks like all the edge cases I found have been noted above in the description. Otherwise this works well, and moves us closer. Going to take a look at the code now.

**Comment by kellyi** [comment]
> the background color of the columns alternates. Was any thought given to this here?

Whoops, missed that one. Should be simple enough to do with index mod 2 though, so I can add a commit to fix it.

Working on the combined hydrology ydomain value, too.

File: src/mmw/js/src/core/chart.js [hunk 3]
```diff
@@ -346,6 +346,8 @@ function renderCompareMultibarChart(chartEl, name, label, colors, stacked, data)
         if (maxValue) {
             yDomainMax = maxValue * 1.2;
         }
+    } else {
+        yDomainMax = precipitation || 1;
```
Comment (kellyi) [inline src/mmw/js/src/core/chart.js]: The fallback's to ensure that if precipitation's set to 0, the graph will still range from 0 to 1.

**Comment by kellyi** [comment]
8a046c3 departs from the wireframes to show the max value on the y-axis:

screen shot 2017-08-10 at 12 14 08 pm

If we decide we *don't* want to show them, I can either drop the commit or we can set `showMaxMin` to be `false` in the future.

**Comment by rajadain** [comment]
0.1 is shown twice in different locations in the Evapotranspiration chart in the screenshot. That can't be right.

**Comment by kellyi** [comment]
> 0.1 is shown twice in different locations in the Evapotranspiration chart in the screenshot. That can't be right.

Ah, looks like we have to set the tick count differently if there are only two values.

**Comment by kellyi** [comment]
I squashed this down to one commit & rebased it on `develop` so it'll be a single commit while adjusting the model & collection hierarchy here.

**Comment by jfrankl** [comment]
Good catch. Let's model this after the proposed loading designs seen here: https://github.com/WikiWatershed/model-my-watershed/issues/1848

While model is running:

1. Add spinner
2. Change opacity of bars in bar graphs

Details in the screenshot below:

mmw-model-running

**Comment by rajadain** [comment]
Closing in favor of #2148.

**Pull request closed without merging**